src/OpenApi/gen/XmlComments/MemberKey.cs

```diff
@@ -44,94 +49,193 @@ public static MemberKey FromMethodSymbol(IMethodSymbol method, Compilation compi
-            MemberType.Type,
-            null,
-            null,
-            null);
+        if (TryGetFormattedTypeName(type, out var typeName))
+        {
+            return new MemberKey(
+                $"typeof({typeName})",
+                MemberType.Type,
+                null,
+                null,
+                null);
+        }
+        return null;
     }
 
     /// Supports replacing generic type arguments to support use of open
     /// generics in `typeof` expressions for the declaring type.
-    private static string ReplaceGenericArguments(string typeName)
+    private static bool TryGetFormattedTypeName(ITypeSymbol typeSymbol, [NotNullWhen(true)] out string? typeName, bool isNestedCall = false)
     {
-        var stack = new Stack<int>();
-        var result = new StringBuilder(typeName);
-        for (var i = 0; i < result.Length; i++)
+        if (typeSymbol is INamedTypeSymbol { OriginalDefinition.SpecialType: SpecialType.System_Nullable_T } nullableType)
         {
-            if (result[i] == '<')
+            typeName = typeSymbol.ToDisplayString(_withTypeParametersFormat);
+            return true;
+        }
+
+        // Handle tuples specially since they are represented as generic
+        // ValueTuple types and trigger the logic for handling generics in
+        // nested values.
+        if (typeSymbol is INamedTypeSymbol { IsTupleType: true } namedType)
+        {
+            return TryHandleTupleType(namedType, out typeName);
+        }
+
+        if (typeSymbol is INamedTypeSymbol { IsGenericType: true } genericType)
+        {
+            // If any of the type arguments are type parameters, then they have not
+            // been substituted for a concrete type and we need to model them as open
+            // generics if possible to avoid emitting a type with type parameters that
+            // cannot be used in a typeof expression.
+            var hasTypeParameters = genericType.TypeArguments.Any(t => t.TypeKind == TypeKind.TypeParameter);
+            var baseTypeName = genericType.ToDisplayString(_sansTypeParametersFormat);
+
+            if (!hasTypeParameters)
+            {
+                var typeArgStrings = new List<string>();
+                var allArgumentsResolved = true;
+
+                // Loop through each type argument to handle nested generics.
+                foreach (var typeArg in genericType.TypeArguments)
+                {
+                    if (TryGetFormattedTypeName(typeArg, out var argTypeName, isNestedCall: true))
+                    {
+                        typeArgStrings.Add(argTypeName);
+                    }
+                    else
+                    {
+                        typeName = null;
+                        return false;
+                    }
+                }
+
+                if (allArgumentsResolved)
+                {
+                    typeName = $"{baseTypeName}<{string.Join(", ", typeArgStrings)}>";
```

Comment:
Not sure what use to use instead, but maybe a different prefix than `base` so someone doesn't think it might refer to inheritance?

Reply:
Updated.